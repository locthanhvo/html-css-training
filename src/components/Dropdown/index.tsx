import { memo, useState } from 'react';
import { Button, Menu, MenuButton, MenuList, MenuItem } from '@chakra-ui/react';

// Icons
import { DropdownIcon } from '@/components/Icons';

interface DropdownProps {
  defaultValue?: string;
  options: { name: string; value: string }[];
  onSelect?: (option: string) => void;
}

const Dropdown = ({ defaultValue, options, onSelect }: DropdownProps) => {
  const [selectedOption, setSelectedOption] = useState(
    defaultValue || options[0].value,
  );

  const handleSelect = (option: string) => {
    setSelectedOption(option);
    onSelect?.(option);
  };

  const optionName = options.find(
    (option) => option.value === selectedOption,
  )?.name;

  return (
    <Menu>
      <MenuButton
        w="516px"
        h="40px"
        textAlign="left"
        fontSize="xs"
        border="1px solid"
        borderColor="inherit"
        as={Button}
        color="gray.400"
        rightIcon={<DropdownIcon />}
      >
        {optionName}
      </MenuButton>
      <MenuList>
        {options.map((option) => (
          <MenuItem
            fontSize="xs"
            color="gray.600"
            key={option.value}
            bgColor={
              selectedOption === option.value ? 'gray.50' : 'transparent'
            }
            onClick={() => handleSelect(option.value)}
          >
            {option.name}
          </MenuItem>
        ))}
      </MenuList>
    </Menu>
  );
};

export default memo(Dropdown);
