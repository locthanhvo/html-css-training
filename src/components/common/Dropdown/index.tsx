import { ChangeEvent, memo, useState, useCallback } from 'react';
import {
  Checkbox,
  CheckboxGroup,
  Menu,
  MenuButton,
  MenuList,
  MenuItem,
  Button,
} from '@chakra-ui/react';
import { ISelectField, ITask } from '@/types';

interface DropdownProps {
  defaultValue?: { name: string; image?: string; value: string }[];
  options: { name: string; image?: string; value: string }[];
  name: string;
  onChange?: (selectField: ISelectField<ITask>) => void;
  field: keyof ITask;
}

const Dropdown = ({
  defaultValue = [],
  options,
  name,
  onChange,
  field,
}: DropdownProps) => {
  const [selectedOptions, setSelectedOptions] =
    useState<{ name: string; image?: string; value: string }[]>(defaultValue);

  const handleSelectOption = useCallback(
    (option: { name: string; image?: string; value: string }) => {
      setSelectedOptions((prevOptions) => {
        const newOptions = prevOptions.some(
          (item) => item.value === option.value,
        )
          ? prevOptions.filter((item) => item.value !== option.value)
          : [...prevOptions, option];
        onChange?.({
          field,
          data: newOptions,
          isError: false,
          onChange: () => {},
        });
        return newOptions;
      });
    },
    [onChange, field],
  );

  const isChecked = useCallback(
    (option: { name: string; image?: string; value: string }) =>
      selectedOptions.some((item) => item.value === option.value),
    [selectedOptions],
  );

  const handleCheckboxClick = (
    e: ChangeEvent<HTMLInputElement>,
    option: { name: string; image?: string; value: string },
  ) => {
    e.stopPropagation();
    handleSelectOption(option);
  };

  return (
    <Menu closeOnSelect={false}>
      <MenuButton
        as={Button}
        w={168}
        border="2px solid"
        borderColor="lightGray"
        color="primary"
        fontSize="md"
      >
        {name}
      </MenuButton>
      <MenuList>
        <CheckboxGroup value={selectedOptions.map((option) => option.value)}>
          {options.map((option) => (
            <MenuItem key={option.value}>
              <Checkbox
                isChecked={isChecked(option)}
                onChange={(e) => handleCheckboxClick(e, option)}
              >
                {option.name}
              </Checkbox>
            </MenuItem>
          ))}
        </CheckboxGroup>
      </MenuList>
    </Menu>
  );
};

export default memo(Dropdown);
