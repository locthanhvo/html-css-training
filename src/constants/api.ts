export const API_BASE_URL = import.meta.env.VITE_BASE_URL

export const API_PATH_URL = {
  USERS_URL: '/users',
}
