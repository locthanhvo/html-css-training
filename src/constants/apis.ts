import { TQueryKey } from '@/types';

export const API_PATHS = {
  USERS: '/users',
  COMPANIES: '/companies',
  BRIGADES: '/brigades',
  CONTROL_TEAMS: '/controls',
  STAGE_COACHES: '/stagecoaches',
  CRCA: '/crca',
};

export const userLoginQueryKeys = {
  USER_LOGIN: 'userLogin',
};

export const userQueryKeys = {
  all: [{ scope: 'users' }] as const,
  lists: () => [{ ...userQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ page, filter }: TQueryKey) =>
    [
      {
        ...userQueryKeys.lists()[0],
        page,
        ...(filter && { filter }),
      },
    ] as const,
  details: () => [{ ...userQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...userQueryKeys.details()[0], id }] as const,
};

export const companyQueryKeys = {
  all: [{ scope: 'companies' }] as const,
  lists: () => [{ ...companyQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ page, filter }: TQueryKey) =>
    [
      {
        ...companyQueryKeys.lists()[0],
        page,
        filter,
      },
    ] as const,
  details: () => [{ ...companyQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...companyQueryKeys.details()[0], id }] as const,
};

export const brigadeQueryKeys = {
  all: [{ scope: 'brigades' }] as const,
  lists: () => [{ ...brigadeQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ filter }: { filter?: Record<string, string> }) =>
    [
      {
        ...brigadeQueryKeys.lists()[0],
        filter,
      },
    ] as const,
  details: () => [{ ...brigadeQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...brigadeQueryKeys.details()[0], id }] as const,
};

export const controlTeamsQueryKeys = {
  all: [{ scope: 'controlTeams' }] as const,
  lists: () => [{ ...controlTeamsQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ filter }: { filter?: Record<string, string> }) =>
    [
      {
        ...controlTeamsQueryKeys.lists()[0],
        filter,
      },
    ] as const,
  details: () =>
    [{ ...controlTeamsQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) =>
    [{ ...controlTeamsQueryKeys.details()[0], id }] as const,
};

export const stageCoachesQueryKeys = {
  all: [{ scope: 'stageCoaches' }] as const,
  lists: () => [{ ...stageCoachesQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ filter }: { filter?: Record<string, string> }) =>
    [
      {
        ...stageCoachesQueryKeys.lists()[0],
        filter,
      },
    ] as const,
  details: () =>
    [{ ...stageCoachesQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) =>
    [{ ...stageCoachesQueryKeys.details()[0], id }] as const,
};

export const crcaQueryKeys = {
  all: [{ scope: 'crca' }] as const,
  lists: () => [{ ...crcaQueryKeys.all[0], entity: 'list' }] as const,
  list: ({ filter }: { filter?: Record<string, string> }) =>
    [
      {
        ...crcaQueryKeys.lists()[0],
        filter,
      },
    ] as const,
  details: () => [{ ...crcaQueryKeys.all[0], entity: 'detail' }] as const,
  detail: (id?: string) => [{ ...crcaQueryKeys.details()[0], id }] as const,
};
