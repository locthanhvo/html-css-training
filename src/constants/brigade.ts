import { TBrigade } from '@/types';

export const INITIAL_BRIGADE: TBrigade = {
  id: '',
  name: '',
  description: '',
  commissioner: '',
  role: '',
  createdAt: '',
  updatedAt: '',
};

export const BRIGADE_ROLE = [
  {
    value: 'admin',
    name: 'Admin',
  },
  {
    value: 'member',
    name: 'Member',
  },
];
