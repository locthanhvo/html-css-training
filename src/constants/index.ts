export * from './router';
export * from './navbar';
export * from './sidebar';
export * from './menu';
export * from './httpsMethod';
export * from './factoryKey';
export * from './message';
export * from './form';
export * from './regex';
export * from './status';
export * from './task';
