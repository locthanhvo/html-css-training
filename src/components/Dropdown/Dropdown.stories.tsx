import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Dropdown from '.';

// Constants
import { BRANCHES } from '@/constants';

export default {
  title: 'Components/Dropdown',
  component: Dropdown,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof Dropdown>;

export const DropdownDefault: Story = {
  args: { options: BRANCHES },
};
