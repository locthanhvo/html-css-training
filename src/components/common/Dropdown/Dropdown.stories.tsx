import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Dropdown from '.';

// Constants
import { LABELS } from '@/constants';

export default {
  title: 'Components/Dropdown',
  component: Dropdown,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta;

type Story = StoryObj<typeof Dropdown>;

export const DropdownDefault: Story = {
  args: { options: LABELS, name: 'Labels' },
};
