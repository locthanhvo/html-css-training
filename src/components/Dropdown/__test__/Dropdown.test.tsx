import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Dropdown from '..';

// Constants
import { BRANCHES } from '@/constants';

describe('Dropdown Component', () => {
  it('Should match snapshot', () => {
    const element = render(
      <BrowserRouter>
        <Dropdown options={BRANCHES} />
      </BrowserRouter>,
    );

    expect(element).toMatchSnapshot();
  });
});
