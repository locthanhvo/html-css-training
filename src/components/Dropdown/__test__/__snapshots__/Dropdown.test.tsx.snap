// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-controls="menu-list-:r1:"
        aria-expanded="false"
        aria-haspopup="menu"
        class="chakra-button chakra-menu__menu-button css-b0lbf4"
        id="menu-button-:r1:"
        type="button"
      >
        <span
          class="css-xl71ch"
        >
          NON VIE
        </span>
        <span
          class="chakra-button__icon css-1hzyiq5"
        >
          <svg
            fill="none"
            height="7"
            viewBox="0 0 12 7"
            width="12"
          >
            <path
              d="M11 1L6 6L1 1"
              stroke="black"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </span>
      </button>
      <div
        class="css-0"
        style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
      >
        <div
          aria-orientation="vertical"
          class="chakra-menu__menu-list css-sjt5nk"
          id="menu-list-:r1:"
          role="menu"
          style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
          tabindex="-1"
        >
          <button
            class="chakra-menu__menuitem css-kjnj8a"
            data-index="0"
            id="menu-list-:r1:-menuitem-:r2:"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            NON VIE
          </button>
          <button
            class="chakra-menu__menuitem css-wntb6v"
            data-index="1"
            id="menu-list-:r1:-menuitem-:r3:"
            role="menuitem"
            tabindex="-1"
            type="button"
          >
            VIE
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <button
      aria-controls="menu-list-:r1:"
      aria-expanded="false"
      aria-haspopup="menu"
      class="chakra-button chakra-menu__menu-button css-b0lbf4"
      id="menu-button-:r1:"
      type="button"
    >
      <span
        class="css-xl71ch"
      >
        NON VIE
      </span>
      <span
        class="chakra-button__icon css-1hzyiq5"
      >
        <svg
          fill="none"
          height="7"
          viewBox="0 0 12 7"
          width="12"
        >
          <path
            d="M11 1L6 6L1 1"
            stroke="black"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </span>
    </button>
    <div
      class="css-0"
      style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
    >
      <div
        aria-orientation="vertical"
        class="chakra-menu__menu-list css-sjt5nk"
        id="menu-list-:r1:"
        role="menu"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
        tabindex="-1"
      >
        <button
          class="chakra-menu__menuitem css-kjnj8a"
          data-index="0"
          id="menu-list-:r1:-menuitem-:r2:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          NON VIE
        </button>
        <button
          class="chakra-menu__menuitem css-wntb6v"
          data-index="1"
          id="menu-list-:r1:-menuitem-:r3:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          VIE
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
